{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0ecf2d4e-d978-4b17-845f-bc85fcef05df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change the working directory\n",
    "import os\n",
    "os.chdir('/home/jovyan/conga/data/test')\n",
    "\n",
    "# Load rpy2\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "af06396a-dd78-448e-993b-c06e3b15e2ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ab_counts: [((1, 1), 7550), ((2, 1), 846), ((1, 2), 223), ((2, 2), 145)]\n",
      "old_unpaired_barcodes: 1106 old_paired_barcodes: 8784 new_stringent_paired_barcodes: 8396 paired_barcode_fraction: 0.9558287795992714\n",
      "compute tcrdist distance matrix for 8039 clonotypes\n",
      "Using C++ TCRdist calculator\n",
      "Read 8039 paired tcrs from file ./outs/merged_clones_AB.dist_50_kpcs_tcrs.tsv\n",
      "making ./outs/merged_clones_AB.dist_50_kpcs_tcrdists.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      ".................................................. 5000\n",
      "..............................\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running KernelPCA with None kernel distance matrix shape= (8039, 8039) D.max()= 504.0 force_Dmax= None\n",
      "writing TCRdist kernel PCs to outfile: ./outs/merged_clones_AB.dist_50_kpcs\n"
     ]
    }
   ],
   "source": [
    "# Run make conga clone\n",
    "# Takes in tcr_meta.csv which points to the filtered contig \n",
    "# annotations files and gives each of them a batch_id.   The resulting file is \n",
    "# merged_clones.tsv.  This script does PCA on that file.\n",
    "\n",
    "%run make_conga_clone.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1b4edb8f-6423-41bb-9244-361e34490413",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──\n",
      "✔ dplyr     1.1.4     ✔ readr     2.1.4\n",
      "✔ forcats   1.0.0     ✔ stringr   1.5.1\n",
      "✔ ggplot2   3.4.4     ✔ tibble    3.2.1\n",
      "✔ lubridate 1.9.3     ✔ tidyr     1.3.0\n",
      "✔ purrr     1.0.2     \n",
      "── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "✖ dplyr::filter() masks stats::filter()\n",
      "✖ dplyr::lag()    masks stats::lag()\n",
      "ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors\n",
      "\n",
      "    WARNING: The R package \"reticulate\" only fixed recently\n",
      "    an issue that caused a segfault when used with rpy2:\n",
      "    https://github.com/rstudio/reticulate/pull/1188\n",
      "    Make sure that you use a version of that package that includes\n",
      "    the fix.\n",
      "    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Attaching SeuratObject\n",
      "\n",
      "R[write to console]: Seurat v4 was just loaded with SeuratObject v5; disabling v5 assays and\n",
      "validation routines, and ensuring assays work in strict v3/v4\n",
      "compatibility mode\n",
      "\n",
      "R[write to console]: Loading required package: SingleCellExperiment\n",
      "\n",
      "R[write to console]: Loading required package: SummarizedExperiment\n",
      "\n",
      "R[write to console]: Loading required package: MatrixGenerics\n",
      "\n",
      "R[write to console]: Loading required package: matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    second, second<-\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:lubridate’:\n",
      "\n",
      "    %within%\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "R[write to console]: Genome matrix has multiple modalities, returning a list of matrices for this genome\n",
      "\n",
      "R[write to console]: Performing log-normalization\n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: Calculating gene variances\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: Calculating feature variances of standardized and clipped values\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: Centering and scaling data matrix\n",
      "\n",
      "  |                                                                            \n",
      "  |                                                                      |   0%\n",
      "  |                                                                            \n",
      "  |===================================                                   |  50%\n",
      "  |                                                                            \n",
      "  |======================================================================| 100%\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: PC_ 1 \n",
      "Positive:  CD3E, IFITM1, LTB, IL32, CD3D, IL7R, TCF7, LDHB, RPS12, PCED1B-AS1 \n",
      "\t   EEF1A1, RPS18, RPS29, CD7, FCMR, LIME1, LEF1, CD27, CCR7, LINC00861 \n",
      "\t   NOSIP, CD2, FLT3LG, CD247, GIMAP5, PRKCQ-AS1, PIK3IP1, BCL11B, MALAT1, ARL4C \n",
      "Negative:  LYZ, CST3, FCN1, S100A9, IFI30, S100A8, VCAN, TYROBP, MNDA, SPI1 \n",
      "\t   SERPINA1, CTSS, LST1, TYMP, CEBPD, AIF1, CD14, FTL, FCER1G, CYBB \n",
      "\t   CSTA, MS4A6A, CSF3R, PSAP, CD68, TNFAIP2, S100A12, LGALS1, NCF2, CFD \n",
      "PC_ 2 \n",
      "Positive:  IL32, CD3E, IFITM1, CD3D, S100A4, IL7R, CD7, ANXA1, S100A10, TCF7 \n",
      "\t   CD247, GIMAP5, S100A6, CD2, LINC00861, SRGN, GZMM, CTSW, LIME1, LDHB \n",
      "\t   ITGB2, VIM, S100A11, NOSIP, LEF1, FLT3LG, PRKCQ-AS1, GAPDH, BCL11B, ARL4C \n",
      "Negative:  MS4A1, CD79A, BANK1, LINC00926, IGHM, NIBAN3, HLA-DQA1, CD79B, FCER2, CD19 \n",
      "\t   FCRLA, HLA-DQB1, SPIB, RALGPS2, HLA-DQA2, FCRL1, CD22, TCL1A, VPREB3, AFF3 \n",
      "\t   HLA-DOB, HLA-DRA, BLK, BLNK, POU2AF1, IGHD, HLA-DPB1, CD24, PAX5, CD40 \n",
      "PC_ 3 \n",
      "Positive:  MPIG6B, CAVIN2, TUBB1, GNG11, PPBP, PRKAR2B, PF4, CCL5, CLU, SPARC \n",
      "\t   GP1BB, GP9, ITGA2B, TREML1, NKG7, GZMB, AC147651.1, PTGS1, CMTM5, MYL9 \n",
      "\t   ITGB3, RAB27B, MMD, GNLY, CST7, KLRD1, PRF1, FGFBP2, TRIM58, TSC22D1 \n",
      "Negative:  RPS12, EEF1A1, RPS18, MALAT1, VIM, LEF1, CCR7, TCF7, LTB, SELL \n",
      "\t   LDHB, RPS29, MAL, NOSIP, TRABD2A, IL7R, JUNB, PIK3IP1, CD3E, CD27 \n",
      "\t   CD3D, OXNAD1, H1FX, AIF1, ARMH1, FOS, LRRN3, PRKCQ-AS1, NELL2, CHRM3-AS2 \n",
      "PC_ 4 \n",
      "Positive:  PRKAR2B, PPBP, CAVIN2, TUBB1, PF4, GNG11, SPARC, CLU, GP1BB, GP9 \n",
      "\t   RGS10, TREML1, MPIG6B, ITGA2B, AC147651.1, CMTM5, PTGS1, NRGN, ITGB3, ACTN1 \n",
      "\t   BEX3, TRIM58, CD9, GMPR, SEPTIN5, MMD, MYL9, CTDSPL, C2orf88, ACRBP \n",
      "Negative:  NKG7, CST7, PRF1, GNLY, KLRD1, GZMB, GZMA, FGFBP2, KLRF1, SPON2 \n",
      "\t   CLIC3, TBX21, FCGR3A, HOPX, SH2D1B, S1PR5, ADGRG1, PTGDR, CCL4, PRSS23 \n",
      "\t   IL2RB, GZMH, MATK, CX3CR1, KIR3DL1, TTC38, CTSW, XCL2, FCRL6, AKR1C3 \n",
      "PC_ 5 \n",
      "Positive:  IFITM3, SELL, S100A12, C20orf204, ANPEP, ACTB, CCR7, CYP1B1, S100A8, RBP7 \n",
      "\t   CD36, SHTN1, AL138963.4, RPS29, CX3CR1, ACTN1, LRRN3, NFE2, CSTA, AIF1 \n",
      "\t   MNDA, LEF1, FCER1G, FGFBP2, CD7, GCA, FPR1, PLBD1, BACH2, AC103591.3 \n",
      "Negative:  NR4A2, NR4A1, JUNB, FOSB, NFKBIA, PPP1R15A, DUSP1, DUSP2, TNFAIP3, BTG2 \n",
      "\t   FOS, JUN, ZFP36, EGR1, UBE2S, ATF3, CD69, DNAJB1, SERTAD1, IL1B \n",
      "\t   TSC22D3, GADD45B, NFKBIZ, CD83, NR4A3, SOCS3, IER5, TUBB4B, CXCL8, ATP2B1-AS1 \n",
      "\n",
      "R[write to console]: Computing nearest neighbor graph\n",
      "\n",
      "R[write to console]: Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 15647\n",
      "Number of edges: 528566\n",
      "\n",
      "Running Louvain algorithm...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Maximum modularity in 10 random starts: 0.8995\n",
      "Number of communities: 22\n",
      "Elapsed time: 2 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Warning:\n",
      "R[write to console]:  The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session\n",
      "\n",
      "R[write to console]: 01:29:08 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "R[write to console]: Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "R[write to console]: Also defined by ‘BiocGenerics’\n",
      "\n",
      "R[write to console]: 01:29:08 Read 15647 rows and found 50 numeric columns\n",
      "\n",
      "R[write to console]: 01:29:08 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "R[write to console]: Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "R[write to console]: Also defined by ‘BiocGenerics’\n",
      "\n",
      "R[write to console]: 01:29:08 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: 01:29:10 Writing NN index file to temp file /tmp/RtmpXZfysn/file4a06c44425c\n",
      "\n",
      "R[write to console]: 01:29:10 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "R[write to console]: 01:29:15 Annoy recall = 100%\n",
      "\n",
      "R[write to console]: 01:29:15 Commencing smooth kNN distance calibration using 1 thread\n",
      "R[write to console]:  with target n_neighbors = 30\n",
      "\n",
      "R[write to console]: 01:29:17 Initializing from normalized Laplacian + noise (using irlba)\n",
      "\n",
      "R[write to console]: 01:29:19 Commencing optimization for 200 epochs, with 705094 positive edges\n",
      "\n",
      "R[write to console]: Using method '\n",
      "R[write to console]: umap\n",
      "R[write to console]: '\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: 0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "R[write to console]: [----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: *\n",
      "R[write to console]: |\n",
      "\n",
      "R[write to console]: 01:29:27 Optimization finished\n",
      "\n",
      "R[write to console]: You created a large dataset with compression and chunking.\n",
      "The chunk size is equal to the dataset dimensions.\n",
      "If you want to read subsets of the dataset, you should testsmaller chunk sizes to improve read times.\n",
      "\n",
      "R[write to console]: You created a large dataset with compression and chunking.\n",
      "The chunk size is equal to the dataset dimensions.\n",
      "If you want to read subsets of the dataset, you should testsmaller chunk sizes to improve read times.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "# Load tidyverse because it's needed for the script but he doesn't load it\n",
    "library(tidyverse)\n",
    "\n",
    "# Rename the output file because it can't be overwritten and will cause an error\n",
    "file.rename(from = \"./outs/merged_counts_matrix.h5\", to = \"./outs/merged_counts_matrix_old.h5\")\n",
    "\n",
    "# Run the make conga happy seurat R script\n",
    "source('make_conga_happy_seurat.R')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8a9aa85f-26ff-4b6a-9c3e-58989081ff26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.10.3\n",
      "scanpy      1.9.6\n",
      "-----\n",
      "PIL                 9.2.0\n",
      "beta_ufunc          NA\n",
      "binom_ufunc         NA\n",
      "bottleneck          1.3.5\n",
      "cffi                1.15.1\n",
      "cloudpickle         2.2.0\n",
      "colorama            0.4.6\n",
      "conga               NA\n",
      "cycler              0.10.0\n",
      "cython_runtime      NA\n",
      "cytoolz             0.12.0\n",
      "dask                2022.11.0\n",
      "dateutil            2.8.2\n",
      "defusedxml          0.7.1\n",
      "exceptiongroup      1.2.0\n",
      "fsspec              2022.11.0\n",
      "google              NA\n",
      "h5py                3.7.0\n",
      "hypergeom_ufunc     NA\n",
      "igraph              0.10.8\n",
      "jinja2              3.1.2\n",
      "joblib              1.2.0\n",
      "kiwisolver          1.4.4\n",
      "leidenalg           0.10.1\n",
      "llvmlite            0.39.1\n",
      "lz4                 4.0.2\n",
      "markupsafe          2.1.1\n",
      "matplotlib          3.6.2\n",
      "mpl_toolkits        NA\n",
      "natsort             8.4.0\n",
      "nbinom_ufunc        NA\n",
      "ncf_ufunc           NA\n",
      "numba               0.56.4\n",
      "numexpr             2.8.3\n",
      "numpy               1.23.5\n",
      "packaging           21.3\n",
      "pandas              1.5.2\n",
      "pkg_resources       NA\n",
      "psutil              5.9.4\n",
      "pyparsing           3.0.9\n",
      "pytz                2022.6\n",
      "ruamel              NA\n",
      "scipy               1.9.3\n",
      "session_info        1.0.0\n",
      "setuptools          65.5.1\n",
      "six                 1.16.0\n",
      "sklearn             1.1.3\n",
      "statsmodels         0.13.5\n",
      "tblib               1.7.0\n",
      "texttable           1.7.0\n",
      "threadpoolctl       3.1.0\n",
      "tlz                 0.12.0\n",
      "toolz               0.12.0\n",
      "typing_extensions   NA\n",
      "wcwidth             0.2.5\n",
      "yaml                6.0\n",
      "-----\n",
      "Python 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:26:04) [GCC 10.4.0]\n",
      "Linux-6.4.16-linuxkit-x86_64-with-glibc2.35\n",
      "-----\n",
      "Session information updated at 2023-12-31 01:31\n",
      "reading: ./outs/merged_counts_matrix.h5 of type 10x_h5\n",
      "total barcodes: 15647 (15647, 19500)\n",
      "reading: ./outs/merged_clones.tsv\n",
      "reading: ./outs/merged_clones_AB.dist_50_kpcs\n",
      "Reducing to the 7943 barcodes (out of 15647) with paired TCR sequence data\n",
      "AnnData object with n_obs × n_vars = 7943 × 19500\n",
      "    obs: 'va', 'ja', 'cdr3a', 'cdr3a_nucseq', 'vb', 'jb', 'cdr3b', 'cdr3b_nucseq'\n",
      "    var: 'gene_ids'\n",
      "    uns: 'conga_results', 'conga_stats', 'organism'\n",
      "    obsm: 'X_pca_tcr'\n",
      "making: merged_conga_test_qc_violins.png\n",
      "making: merged_conga_test_counts_vs_genes.png\n",
      "filtered out 0 cells with more than 3000 genes\n",
      "filtered out 0 cells with more than 0.1 percent mito\n",
      "unable to find feature_types varname\n",
      "Index(['AL627309.1', 'AL627309.5', 'LINC01409', 'LINC01128', 'LINC00115',\n",
      "       'FAM41C', 'NOC2L', 'KLHL17', 'PLEKHN1', 'HES4',\n",
      "       ...\n",
      "       'HORMAD2-AS1', 'TMPRSS6', 'AL022314.1', 'C22orf23', 'AL022328.1',\n",
      "       'ARMCX3-AS1', 'TCEAL2', 'RHOXF1', 'GPC3', 'BX004987.1'],\n",
      "      dtype='object', length=17420)\n",
      "excluding 97 TR genes (89 variable)\n",
      "excluding 10 sexlinked genes\n",
      "total of 816 variable genes (7943, 17420)\n",
      "Normalize and logging matrix...\n",
      "unable to find feature_types varname\n",
      "Index(['LINC00115', 'ISG15', 'TNFRSF18', 'TNFRSF4', 'MMP23B', 'PHF13', 'RBP7',\n",
      "       'AL109811.2', 'AL021155.5', 'TNFRSF1B',\n",
      "       ...\n",
      "       'CD40LG', 'FLNA', 'MPP1', 'F8A1', 'MT-ND1', 'MT-ND2', 'MT-ATP8',\n",
      "       'MT-ND4', 'MT-ND6', 'KRT1'],\n",
      "      dtype='object', length=816)\n",
      "normalize_and_log_the_raw_matrix: adata.raw.X.max()=  611.0\n",
      "run reduce_to_single_cell_per_clone\n",
      "compute pca to find rep cell for each clone (7943, 816)\n",
      "num_clones: 7629\n",
      "normalize_and_log_the_raw_matrix:: matrix is already logged\n",
      "choose representative cell for clone: 0 7629 (7943, 816)\n",
      "choose representative cell for clone: 1000 7629 (7943, 816)\n",
      "choose representative cell for clone: 2000 7629 (7943, 816)\n",
      "choose representative cell for clone: 3000 7629 (7943, 816)\n",
      "choose representative cell for clone: 4000 7629 (7943, 816)\n",
      "choose representative cell for clone: 5000 7629 (7943, 816)\n",
      "choose representative cell for clone: 6000 7629 (7943, 816)\n",
      "choose representative cell for clone: 7000 7629 (7943, 816)\n",
      "reduce from 7943 cells to 7629 cells (one per clonotype)\n",
      "run cluster_and_tsne_and_umap\n",
      "computing X_pca_gex using sc.tl.pca\n",
      "OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.\n",
      "running 1D UMAP gex\n",
      "ran leiden clustering: leiden_gex\n",
      "running 1D UMAP tcr\n",
      "ran leiden clustering: leiden_tcr\n",
      "compute D gex 7629\n",
      "argpartitions: 0.01 7629 gex\n",
      "calculate nndists: gex 0.01\n",
      "DONE calculating nndists: gex 0.01\n",
      "argpartitions: 0.1 7629 gex\n",
      "compute D tcr 7629\n",
      "argpartitions: 0.01 7629 tcr\n",
      "calculate nndists: tcr 0.01\n",
      "DONE calculating nndists: tcr 0.01\n",
      "argpartitions: 0.1 7629 tcr\n",
      "setup_tcr_cluster_names: ['0_AV8', '1', '2', '3_AV14++', '4', '5_AV12', '6', '7_BV20', '8_AV9', '9_AV26', '10_AV21', '11_AV1', '12_BV30', '13_BV20', '14_BV18', '15', '16_BV20', '17_BV20', '18_AV1']\n",
      "find_neighbor_neighbor_interactions:\n",
      "find_neighbor_cluster_interactions:\n",
      "find_neighbor_cluster_interactions:\n",
      "find_neighbor_neighbor_interactions:\n",
      "find_neighbor_cluster_interactions:\n",
      "find_neighbor_cluster_interactions:\n",
      "saved graph_vs_graph results to tsvfile: merged_conga_test_graph_vs_graph.tsv\n",
      "making tcr score table, #features= 35\n",
      "run rank_genes_groups [('none', 7256), ('clp_2_3', 104), ('clp_9_11', 101), ('clp_9_18', 29), ('clp_2_5', 26), ('clp_2_11', 23), ('clp_2_2', 23), ('clp_0_0', 23), ('clp_6_3', 13), ('clp_2_1', 13), ('clp_6_5', 10), ('clp_2_6', 8)]\n",
      "making cluster logos: 0 11 merged_conga_test_graph_vs_graph_logos.png\n",
      "making cluster logos: 1 11 merged_conga_test_graph_vs_graph_logos.png\n",
      "making cluster logos: 2 11 merged_conga_test_graph_vs_graph_logos.png\n",
      "making cluster logos: 3 11 merged_conga_test_graph_vs_graph_logos.png\n",
      "making cluster logos: 4 11 merged_conga_test_graph_vs_graph_logos.png\n",
      "making cluster logos: 5 11 merged_conga_test_graph_vs_graph_logos.png\n",
      "making cluster logos: 6 11 merged_conga_test_graph_vs_graph_logos.png\n",
      "making cluster logos: 7 11 merged_conga_test_graph_vs_graph_logos.png\n",
      "making cluster logos: 8 11 merged_conga_test_graph_vs_graph_logos.png\n",
      "making cluster logos: 9 11 merged_conga_test_graph_vs_graph_logos.png\n",
      "making cluster logos: 10 11 merged_conga_test_graph_vs_graph_logos.png\n",
      "making: merged_conga_test_graph_vs_graph_logos.png\n",
      "writing help message to file: merged_conga_test_graph_vs_graph_logos.png_README.txt\n",
      "gex_indegree_bias: nbr_frac= 0.0100 DescribeResult(nobs=7629, minmax=(0.0, 10.697368421052632), mean=1.0, variance=1.1504959857762698, skewness=2.5635033214283487, kurtosis=9.875902123884385)\n",
      "tcr_indegree_bias: nbr_frac= 0.0100 DescribeResult(nobs=7629, minmax=(0.09210526315789473, 7.4868421052631575), mean=1.0, variance=0.313452715393208, skewness=1.9838267000401513, kurtosis=8.30776613495199)\n",
      "indegree_bias_correlation: nbr_frac= 0.0100 LinregressResult(slope=-0.0026038230754419772, intercept=1.002603823075442, rvalue=-0.004988477378020771, pvalue=0.6630938700154176, stderr=0.005976696289892416, intercept_stderr=0.00876426049420189)\n",
      "gex_indegree_bias: nbr_frac= 0.1000 DescribeResult(nobs=7629, minmax=(0.0, 5.562992125984252), mean=1.0, variance=0.7533395369039761, skewness=1.4356654901649193, kurtosis=2.3447380483786002)\n",
      "tcr_indegree_bias: nbr_frac= 0.1000 DescribeResult(nobs=7629, minmax=(0.1010498687664042, 5.784776902887139), mean=1.0, variance=0.33059019044391286, skewness=2.05248374902387, kurtosis=7.279065841375061)\n",
      "indegree_bias_correlation: nbr_frac= 0.1000 LinregressResult(slope=0.005965455874274483, intercept=0.9940345441257256, rvalue=0.009005213975492083, pvalue=0.4316098791560845, stderr=0.007584991837056321, intercept_stderr=0.010043287634384977)\n",
      "making: merged_conga_test_results_summary.html\n",
      "make_html_summary:: no results for content_tag graph_vs_graph_stats\n",
      "make_html_summary:: no results for content_tag tcr_clumping\n",
      "make_html_summary:: no results for content_tag tcr_clumping_logos\n",
      "make_html_summary:: no results for content_tag tcr_db_match\n",
      "make_html_summary:: no results for content_tag tcr_db_match_plot\n",
      "make_html_summary:: no results for content_tag tcr_graph_vs_gex_features\n",
      "make_html_summary:: no results for content_tag tcr_graph_vs_gex_features_plot\n",
      "make_html_summary:: no results for content_tag tcr_graph_vs_gex_features_panels\n",
      "make_html_summary:: no results for content_tag tcr_genes_vs_gex_features\n",
      "make_html_summary:: no results for content_tag tcr_genes_vs_gex_features_panels\n",
      "make_html_summary:: no results for content_tag gex_graph_vs_tcr_features\n",
      "make_html_summary:: no results for content_tag gex_graph_vs_tcr_features_plot\n",
      "make_html_summary:: no results for content_tag gex_graph_vs_tcr_features_panels\n",
      "make_html_summary:: no results for content_tag graph_vs_features_gex_clustermap\n",
      "make_html_summary:: no results for content_tag graph_vs_features_tcr_clustermap\n",
      "make_html_summary:: no results for content_tag graph_vs_summary\n",
      "make_html_summary:: no results for content_tag gex_clusters_tcrdist_trees\n",
      "make_html_summary:: no results for content_tag conga_threshold_tcrdist_tree\n",
      "make_html_summary:: no results for content_tag hotspot_features\n",
      "make_html_summary:: no results for content_tag hotspot_gex_umap\n",
      "make_html_summary:: no results for content_tag hotspot_gex_clustermap\n",
      "make_html_summary:: no results for content_tag hotspot_tcr_umap\n",
      "make_html_summary:: no results for content_tag hotspot_tcr_clustermap\n",
      "make_html_summary:: no results for content_tag batch_umaps\n",
      "DONE\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "# Finally run conga\n",
    "\n",
    "!./run_conga.sh"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,Rmd"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
